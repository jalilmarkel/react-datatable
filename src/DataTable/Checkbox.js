import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { handleFunctionProps } from './util';

const baseCheckboxStyle = { fontSize: '18px', cursor: 'pointer' };

export default class Checkbox extends PureComponent {
  static propTypes = {
    indeterminate: PropTypes.bool,
    component: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.func,
    ]),
    onClick: PropTypes.func.isRequired,
    componentOptions: PropTypes.object,
    data: PropTypes.object,
    index: PropTypes.number,
    style: PropTypes.object,
  };

  static defaultProps = {
    indeterminate: false,
    component: 'input',
    componentOptions: {
      style: {},
    },
    data: {},
    index: null,
    style: null,
  };

  componentDidMount() {
    const { indeterminate } = this.props;

    this.el.indeterminate = indeterminate;
  }

  componentDidUpdate(prevProps) {
    const { indeterminate } = this.props;

    if (prevProps.indeterminate !== indeterminate) {
      this.el.indeterminate = indeterminate;
    }
  }

  handleClick = e => {
    const { onClick, data, index } = this.props;

    onClick(data, index, e);
  }

  render() {
    // remove indeterminate to prevent browser warnings
    const {
      component,
      componentOptions,
      indeterminate,
      ...rest
    } = this.props;

    const TagName = component || 'input';
    const baseStyle = TagName !== 'input' ? componentOptions.style : baseCheckboxStyle;
    const resolvedComponentOptions = handleFunctionProps(componentOptions, indeterminate);

    return (
      <TagName
        type="checkbox"
        {...rest}
        {...resolvedComponentOptions}
        // allow this component to fully control these options
        ref={el => { this.el = el; }}
        style={baseStyle}
        onClick={this.handleClick}
        onChange={() => null} // prevent uncontrolled checkbox warnings -  we don't need onChange
      />
    );
  }
}
