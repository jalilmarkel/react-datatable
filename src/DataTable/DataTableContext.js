import React, { PureComponent, createContext } from 'react';
import PropTypes from 'prop-types';

const defaultState = {
  columns: [],
  rows: [],
  selectedRows: [],
  firstCellIndex: 0,
};

const DataTableContext = createContext(defaultState);

export class DataTableProvider extends PureComponent {
  static propTypes = {
    initialState: PropTypes.object.isRequired,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]).isRequired,
  };

  render() {
    const { children, initialState } = this.props;

    return (
      <DataTableContext.Provider value={initialState}>
        {children}
      </DataTableContext.Provider>
    );
  }
}

export const DataTableConsumer = DataTableContext.Consumer;
