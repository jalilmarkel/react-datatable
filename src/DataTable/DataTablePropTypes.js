import React from 'react';
import PropTypes from 'prop-types';

export const propTypes = {
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  selectableRows: PropTypes.bool,
  expandableRows: PropTypes.bool,
  keyField: PropTypes.string,
  progressPending: PropTypes.bool,
  progressComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  progressCentered: PropTypes.bool,
  expanderStateField: PropTypes.string,
  expandableRowsComponent: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]),
  selectableRowsComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  selectableRowsComponentProps: PropTypes.object,
  customTheme: PropTypes.object,
  sortIcon: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  striped: PropTypes.bool,
  highlightOnHover: PropTypes.bool,
  pointerOnHover: PropTypes.bool,
  onServerSort: PropTypes.func,
  actions: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  contextTitle: PropTypes.string,
  contextActions: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  onTableUpdate: PropTypes.func,
  clearSelectedRows: PropTypes.bool,
  defaultSortField: PropTypes.string,
  defaultSortAsc: PropTypes.bool,
  columns: PropTypes.array,
  data: PropTypes.array,
  className: PropTypes.string,
  style: PropTypes.object,
  responsive: PropTypes.bool,
  overflowY: PropTypes.bool,
  overflowYOffset: PropTypes.string,
  noDataComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  disabled: PropTypes.bool,
  noHeader: PropTypes.bool,
  onRowClicked: PropTypes.func,
  fixedHeader: PropTypes.bool,
};

export const defaultProps = {
  title: '',
  keyField: 'id',
  selectableRows: false,
  expandableRows: false,
  progressPending: false,
  progressComponent: <h2>Loading...</h2>,
  progressCentered: false,
  expanderStateField: '$$expander',
  expandableRowsComponent: <div>Add a custom expander component. Use props.data for row data</div>,
  selectableRowsComponent: 'input',
  selectableRowsComponentProps: {},
  customTheme: {},
  sortIcon: false,
  striped: false,
  highlightOnHover: false,
  pointerOnHover: false,
  onServerSort: null,
  contextTitle: '',
  contextActions: [],
  onTableUpdate: null,
  clearSelectedRows: false,
  defaultSortField: null,
  defaultSortAsc: true,
  columns: [],
  data: [],
  className: null,
  style: {},
  responsive: true,
  overflowY: false,
  overflowYOffset: '250px',
  noDataComponent: 'There are no records to display',
  disabled: false,
  noHeader: false,
  onRowClicked: null,
  fixedHeader: false,
};
