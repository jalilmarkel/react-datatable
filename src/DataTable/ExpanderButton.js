import React from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

const ButtonStyle = styled.button`
  outline: none;
  border: none;
  display: block;
  width: 42px;
  height: 42px;
  background-color: transparent;
  background-image: url(${props => props.theme.expander[props.expanded ? 'expandedButton' : 'collapsedButton']});
  background-position: center center;
  background-repeat: no-repeat;

  &:hover {
    cursor: pointer;
  }
`;

const ExpanderButton = ({ expanded, children, data, index, onToggled }) => {
  const handleToggle = e =>
    onToggled && onToggled(data, index, e);

  return (
    <ButtonStyle
      onClick={handleToggle}
      expanded={expanded}
    >
      {children}
    </ButtonStyle>
  );
};

ExpanderButton.propTypes = {
  data: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  expanded: PropTypes.bool,
  onToggled: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

ExpanderButton.defaultProps = {
  onToggled: null,
  children: null,
  expanded: false,
};

export default withTheme(ExpanderButton);
