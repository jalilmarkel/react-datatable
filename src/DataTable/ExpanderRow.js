import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

// Make "data" available on our any child component
// eslint-disable-next-line arrow-body-style
const renderChildren = (children, data) => {
  return Children.map(children, child => cloneElement(child, { data }));
};

const ExpanderRowStyle = styled.div`
  width: 100%;
  box-sizing: border-box;
  background-color: ${props => props.theme.expander.backgroundColor};
`;

const ExpanderRow = ({
  data,
  children,
}) => (
  <ExpanderRowStyle>
    {renderChildren(children, data)}
  </ExpanderRowStyle>
);

ExpanderRow.propTypes = {
  data: PropTypes.object,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

ExpanderRow.defaultProps = {
  data: {},
  children: null,
};

export default withTheme(ExpanderRow);
