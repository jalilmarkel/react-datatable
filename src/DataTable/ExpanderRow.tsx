import * as React from 'react';
import styled, { CSSObject } from 'styled-components';
import { ComponentProps, ExpandableRowsComponent } from './types';

const ExpanderRowStyle = styled.div<{
	extendedRowStyle: CSSObject;
}>`
	width: 100%;
	box-sizing: border-box;
	${({ theme }) => theme.expanderRow.style};
	${({ extendedRowStyle }) => extendedRowStyle};
`;

type ExpanderRowProps<T> = {
	data: T;
	component: ExpandableRowsComponent;
	extendedRowStyle: CSSObject;
	extendedClassNames: string;
	componentProps: ComponentProps;
};

function ExpanderRow<T>({
	data,
	component,
	componentProps,
	extendedRowStyle,
	extendedClassNames,
}: ExpanderRowProps<T>): JSX.Element {
	const ExpandableComponent = component;
	// we need to strip of rdt_TableRow from extendedClassNames
	const classNamesSplit = extendedClassNames.split(' ').filter(c => c !== 'rdt_TableRow');
	const classNames = ['rdt_ExpanderRow', ...classNamesSplit].join(' ');

	return (
		<ExpanderRowStyle className={classNames} extendedRowStyle={extendedRowStyle}>
			<ExpandableComponent data={data} {...componentProps} />
		</ExpanderRowStyle>
	);
}

export default React.memo(ExpanderRow);
