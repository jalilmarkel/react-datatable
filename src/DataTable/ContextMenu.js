import React from 'react';
import styled from 'styled-components';
import { DataTableConsumer } from './DataTableContext';

const Title = styled.div`
  color: ${props => props.theme.header.contextMenu.fontColor};
  font-size: ${props => props.theme.header.contextMenu.fontSize};
  font-weight: 400;
`;

const ContextMenuStyle = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: ${props => props.theme.header.contextMenu.backgroundColor};
  z-index: 1;
  transform: ${props => (props.visible ? 'translate3d(0, 0, 0)' : 'translate3d(0, -100%, 0)')};
  transition-duration: ${props => props.theme.header.contextMenu.transitionTime};
  transition-property: transform;
  align-items: center;
  justify-content: space-between;
  display: flex;
  padding: 16px 16px 16px 24px;
`;

const generateDefaultContextTitle = (contextTitle, selectedCount) => {
  if (selectedCount === 0) {
    return null;
  }

  return contextTitle || `${selectedCount} item${selectedCount > 1 ? 's' : ''} selected`;
};

const ContextMenu = () => (
  <DataTableConsumer>
    {({ contextTitle, contextActions, selectedCount }) => (
      <ContextMenuStyle visible={selectedCount > 0}>
        <Title>
          {generateDefaultContextTitle(contextTitle, selectedCount)}
        </Title>

        <div>
          {contextActions}
        </div>
      </ContextMenuStyle>
    )}
  </DataTableConsumer>
);

export default ContextMenu;
