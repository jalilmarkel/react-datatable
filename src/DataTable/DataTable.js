import React, { Component } from 'react';
import { ThemeProvider } from 'styled-components';
import orderBy from 'lodash/orderBy';
import isEqual from 'lodash/isEqual';
import merge from 'lodash/merge';
import { polyfill } from 'react-lifecycles-compat';
import Table from './Table';
import TableHead from './TableHead';
import TableHeadRow from './TableHeadRow';
import TableRow from './TableRow';
import TableCol from './TableCol';
import TableColCheckbox from './TableColCheckbox';
import ExpanderRow from './ExpanderRow';
import TableHeader from './TableHeader';
import TableBody from './TableBody';
import ResponsiveWrapper from './ResponsiveWrapper';
import ProgressWrapper from './ProgressWrapper';
import TableWrapper from './TableWrapper';
import NoData from './NoData';
import { propTypes, defaultProps } from './DataTablePropTypes';
import { decorateColumns, getSortDirection, calcFirstCellIndex } from './util';
import { handleSelectAll, handleRowChecked, toggleExpand, handleSort, clearSelected } from './statemgmt';
import defaultTheme from '../themes/default';

const recalculateRows = ({ defaultSortField, data }, { sortDirection }) =>
  (defaultSortField ? orderBy(data, defaultSortField, sortDirection) : data);

class DataTable extends Component {
  static propTypes = propTypes;
  static defaultProps = defaultProps;

  static getDerivedStateFromProps(props, state) {
    // allow clearing of rows via passed clearSelectedRows toggle prop
    if (props.clearSelectedRows !== state.clearSelectedRows) {
      return clearSelected(props.clearSelectedRows);
    }
    // Keep data state in sync if it changes
    if (!isEqual(props.data, state.data) || !state.rows.length) {
      return {
        data: props.data,
        rows: recalculateRows(props, state),
      };
    }

    if (props.defaultSortAsc !== state.defaultSortAsc
      || props.defaultSortField !== state.defaultSortField) {
      return {
        sortDirection: getSortDirection(props.defaultSortAsc),
        sortColumn: props.defaultSortField,
      };
    }

    return null;
  }

  constructor(props) {
    super(props);

    const sortDirection = getSortDirection(props.defaultSortAsc);
    this.columns = decorateColumns(props.columns);
    this.state = {
      allSelected: false,
      selectedCount: 0,
      selectedRows: [],
      sortColumn: props.defaultSortField,
      sortDirection,
      rows: [],
      clearSelectedRows: false,
      // eslint-disable-next-line react/no-unused-state
      defaultSortAsc: props.defaultSortAsc,
      // eslint-disable-next-line react/no-unused-state
      defaultSortField: props.defaultSortField,
      // eslint-disable-next-line react/no-unused-state
      data: props.data,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.onTableUpdate &&
      (prevState.selectedRows !== this.state.selectedRows
      || prevState.sortDirection !== this.state.sortDirection
      || prevState.sortColumn !== this.state.sortColumn)
    ) {
      const { allSelected, selectedCount, selectedRows, rows, sortColumn, sortDirection, clearSelectedRows } = this.state;

      this.props.onTableUpdate({
        allSelected,
        selectedCount,
        selectedRows,
        rows,
        sortColumn,
        sortDirection,
        clearSelectedRows,
      });
    }
  }

  generateDefaultContextTitle() {
    const { contextTitle } = this.props;
    const { selectedCount } = this.state;

    return contextTitle || `${selectedCount} item${selectedCount > 1 ? 's' : ''} selected`;
  }

  handleSelectAll = () => {
    this.setState(state => handleSelectAll(state));
  }

  handleRowChecked = row => {
    this.setState(state => handleRowChecked(row, state));
  }

  handleRowClicked = (row, index, e) => {
    if (this.props.onRowClicked) {
      this.props.onRowClicked(row, index, e);
    }
  }

  toggleExpand = row => {
    this.setState((state, props) => toggleExpand(props, row, state));
  }

  handleSort = ({ selector, sortable }) => {
    this.setState((state, props) => handleSort(props, selector, sortable, state));

    if (sortable && this.props.onServerSort) {
      this.props.onServerSort(this.state.sortColumn, this.state.sortDirection);
    }
  }

  renderColumns() {
    const {
      selectableRows,
      expandableRows,
      sortIcon,
    } = this.props;

    const {
      sortColumn,
      sortDirection,
    } = this.state;

    const firstCellIndex = calcFirstCellIndex(selectableRows, expandableRows);

    return (
      this.columns.map(column => (
        <TableCol
          key={column.id}
          column={column}
          onColumnClick={this.handleSort}
          sortField={sortColumn}
          sortDirection={sortDirection}
          sortIcon={sortIcon}
          firstCellIndex={firstCellIndex}
        />))
    );
  }

  renderRows() {
    const {
      selectableRows,
      expandableRows,
      expandableRowsComponent,
      expanderStateField,
      striped,
      highlightOnHover,
      keyField,
      pointerOnHover,
      selectableRowsComponent,
      selectableRowsComponentProps,
      onRowClicked,
    } = this.props;

    const {
      rows,
    } = this.state;
    const getExpanderRowbByParentId = parent => rows.find(r => r.id === parent);
    const firstCellIndex = calcFirstCellIndex(selectableRows, expandableRows);

    return (
      rows.map((row, index) => {
        if (row[expanderStateField]) {
          return (
            <ExpanderRow
              key={`expander--${row[keyField]}`}
              data={getExpanderRowbByParentId(row.parent)}
            >
              {expandableRowsComponent}
            </ExpanderRow>
          );
        }

        return (
          <TableRow
            key={row[this.props.keyField] || index}
            striped={striped}
            highlightOnHover={highlightOnHover}
            pointerOnHover={pointerOnHover}
            columns={this.columns}
            rows={this.state.rows}
            row={row}
            index={index}
            keyField={keyField}
            onRowClicked={this.handleRowClicked}
            rowClickable={!!onRowClicked}
            checkboxComponent={selectableRowsComponent}
            checkboxComponentOptions={selectableRowsComponentProps}
            onRowSelected={this.handleRowChecked}
            selectableRows={selectableRows}
            selectedRows={this.state.selectedRows}
            expandableRows={expandableRows}
            onToggled={this.toggleExpand}
            firstCellIndex={firstCellIndex}
          />
        );
      })
    );
  }

  renderTableHead() {
    const {
      selectableRows,
      selectableRowsComponent,
      selectableRowsComponentProps,
      expandableRows,
    } = this.props;

    const {
      selectedRows,
      allSelected,
    } = this.state;

    const isIndeterminate = selectedRows.length > 0 && !allSelected;

    return (
      <TableHead>
        <TableHeadRow>
          {selectableRows &&
            <TableColCheckbox
              onClick={this.handleSelectAll}
              checked={allSelected}
              checkboxComponent={selectableRowsComponent}
              checkboxComponentOptions={selectableRowsComponentProps}
              indeterminate={isIndeterminate}
            />}
          {expandableRows && <div style={{ width: '42px' }} />}
          {this.renderColumns()}
        </TableHeadRow>
      </TableHead>
    );
  }

  render() {
    const {
      title,
      customTheme,
      actions,
      contextActions,
      className,
      style,
      responsive,
      overflowY,
      overflowYOffset,
      progressPending,
      progressComponent,
      progressCentered,
      noDataComponent,
      disabled,
      noHeader,
      fixedHeader,
    } = this.props;

    const {
      rows,
      selectedCount,
    } = this.state;

    const theme = merge(defaultTheme, customTheme);

    return (
      <ThemeProvider theme={theme}>
        <ResponsiveWrapper
          responsive={responsive}
          className={className}
          style={style}
          overflowYOffset={overflowYOffset}
          overflowY={overflowY}
        >
          {!noHeader &&
          <TableHeader
            title={title}
            showContextMenu={selectedCount > 0}
            contextTitle={this.generateDefaultContextTitle()}
            contextActions={contextActions}
            actions={actions}
            pending={progressPending}
          />}

          <TableWrapper>
            {progressPending &&
              <ProgressWrapper
                component={progressComponent}
                centered={progressCentered}
              />}

            {!rows.length > 0 && !progressPending &&
              <NoData component={noDataComponent} />}

            {rows.length > 0 &&
              <Table disabled={disabled}>
                {this.renderTableHead()}

                <TableBody
                  fixedHeader={fixedHeader}
                  hasOffset={overflowY}
                  offset={overflowYOffset}
                >
                  {this.renderRows()}
                </TableBody>
              </Table>}
          </TableWrapper>
        </ResponsiveWrapper>
      </ThemeProvider>
    );
  }
}

export default polyfill(DataTable);
