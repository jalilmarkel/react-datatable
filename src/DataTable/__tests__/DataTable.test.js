import 'jest-styled-components';

import React from 'react';
import { shallow } from 'enzyme';

import DataTable from '../DataTable';

test('component <DataTable /> should render correctly', () => {
  const wrapper = shallow(<DataTable data={[]} columns={[]} />);

  expect(wrapper.dive().dive()).toMatchSnapshot();
});
