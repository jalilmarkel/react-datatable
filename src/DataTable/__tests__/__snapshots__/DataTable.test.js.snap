// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <DataTable /> should render correctly 1`] = `
.c0 {
  position: relative;
  overflow-x: auto;
}

<div
  className="c0"
  style={Object {}}
>
  <WithTheme(TableHeader)
    contextActions={Array []}
    contextTitle="0 item selected"
    showContextMenu={false}
    title=""
  />
  <TableWrapper>
    <NoDataWrapper
      component="There are no records to display"
    />
  </TableWrapper>
</div>
`;
